README.md
oracle-awr-report.py is easy-to-use data exporter from different data sources (Oracle Database, ssh) to xlsx (Excel)

# Key features
* Export data from Oracle database to xlsx worksheet
* Export data by shell script to xlsx worksheet
* Charts drawing
* Customizing charts
* Adding computed columns

AWR SQL queries by [MaksimIvanovPerm](https://github.com/MaksimIvanovPerm)

# Prerequisite
* python3
* cx_Oracle
* xlswriter
* paramiko
* scp
* logging
* coloredlogs

# Usage

```
$ python3 oracle-awr-report.py 
```

## Main configuration **conf.d/general.conf**
_All config files openes with utf-8 encoding_

## Report configuration file
Format:
```
id:sql script:chart title:chart config file:columns config file
id:plugin:source file:worksheet title:chart title:charts config file:columns config file
```
where charts config file is one of following:
1. **none** - not print chart
2. **default** - use all data to print one chart only
3. **custom charts configuration file name**

where columns config file is one of following:
1. **default** - not use custom columns
2. **custom columns configuration file name**

Example:
```
1:oradb:oradb/RDBMSServiceTime.sql:RDBMSServiceTime:DB service time structure:default:default
2:oradb:oradb/WaitTimeStructure.sql:WaitTimeStructure:DB wait time structure:default:default
3:oradb:oradb/RedoStat.sql:RedoStat::conf.d/redostat.conf:default
4:ssh:ssh/atop_parser.sh --loglevel SILENT2 --swap:atop_swap:atopSWP:default:default
5:ssh:ssh/atop_parser.sh --loglevel SILENT2 --dsk sdc:atop_sdc::conf.d/atop_dsk_charts.conf:conf.d/atop_dsk_columns.conf
```

## Custom charts configuration file
Format:
```
# id:column number in char:column title:chart type

```

where chart type is one of following:
1. **line**
2. **scatter**

Example:
```
1:25:Redo writes per hour:line
1:27::
2:34:Redo write latency, ms:line
3:27:Redo write structure:line
3:33::
```

## Custom columns configuration file
Format:
```
column number:column title:formula
```

**\_ROWID\_** - is service variable for rowid substitution

Example:
```
11:db_file_sequential_read:=UserIOWaitsTime!W_ROWID_/10000
12:userio:=WaitTimeStructure!D_ROWID_/10000
13:rr-latency:=C_ROWID_/E_ROWID_
```

## Report worksheet example
![DB wait time structure](DOC/oracle-awr-report.png)

# Plugins

## ORADB
Plugin for gathering data from Oracle Database

## SSH
Plugin for gathering data by shell script via ssh.

Shell script must output one line with format:
```
File_with_data unixtimestamp column2_title column2_title columnt3_title...
```

Example:
```
/tmp/atop_parser_CPU_20190424_2018_24444.log unixtimestamp sys user irq idle wait
```

So, first column in data file must be unixtimestamp!

# ATOP
For gathering atop data from host prepared shell script **atop_parser.sh**. It work with atop log files /var/log/atop/atop_* created with interval=600 seconds

***Note:** atop device name (in atop lvm statistic) has 12 chars length, so in long lvm names you should use, as parameter of lvm name for atop_parser.sh, tail of logical volume name*



CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

## [unreleased]
### Changed
- fixed RedoStat query
- fixed atop dsk columns config
- improved logging
- added atop_parser.sh lvm support
- added atop_parser.sh per user cpu/memory statistic support

## [1.1.0] - 2019.04.24
### Changed
- added plugin support
- added SSH plugin
- added shell script atop_parser.sh for gather atop data
- added scatter charts support for custom charts
- removed command line arguments support

## [1.0.1] - 2018.10.09
### Changed
- added date to workbook file name
- added dbname to custom charts titles
- fixed bug in DEBUG logging
- added configuration parameters HOST,PORT,SID,SERVICE_NAME as an alternative to TNS_ALIAS
- added reading password from command line, when it not preset in config file
- fixed opening config files with utf-8 encoding
- added command line arguments support

## [1.0.0] - 2018-09-24
###
- Initial release of oracle awr report python script


oracle-awr-report.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
import logging
import coloredlogs
import xlsxwriter

from datetime import date
from importlib import import_module

from plugins.plugin_xlsx import *
from plugins.plugin_report_conf import *
# from plugins.plugin_oradb import *
from plugins import *

# __all__

# variables definition
v_configuration = {}
#
v_parameter_name = ''
v_parameter_value = ''
v_prefix = ''
v_worksheet_name = ''
#
v_dbid = int()
v_begin_snap = int()
v_end_snap = int()
v_log_level = 30
#
v_column_title = []
v_conf = []
v_files = []

# ==================================================================
# ==================================================================
# Main routine
# ==================================================================
# ==================================================================

os.environ["NLS_LANG"] = "AMERICAN_AMERICA.AL32UTF8"

v_configuration["GENERAL_CONF"] = "conf.d/general.conf"

logging.basicConfig(level=logging.WARNING,
                    format='%(asctime)s %(hostname)s %(name)s[%(process)d] %(levelname)s %(message)s')

# if using GENERAL_CONF, open it and parse
if v_configuration.get("GENERAL_CONF"):
    try:
        # try to open main configuration file
        if sys.version_info.major >= 3:
            v_config_file = open(v_configuration["GENERAL_CONF"], encoding="utf_8")
        if sys.version_info.major == 2:
            v_config_file = open(v_configuration["GENERAL_CONF"])
    except OSError as err:
        logging.critical("OS error: {0}".format(err))
        raise
    except Exception:
        logging.critical("Unexpected error:", sys.exc_info()[0])
        raise
    # for all rows in file
    for v_text_string in v_config_file.readlines():
        # if row is not empty and don't start from #
        if not v_text_string.startswith("\n") and not v_text_string.startswith("#"):
            v_parameter_name, v_parameter_value = v_text_string.split('=')
            v_configuration[v_parameter_name] = v_parameter_value.replace("\r", "").replace("\n", "").replace("\"", "")

# LOGGING
v_logger = logging.getLogger("oracle-awr-export")
if not v_configuration.get("LOGGING"):
    v_log_level = 30
    v_configuration["LOGGING"]="WARNING"
elif v_configuration["LOGGING"].upper() == "DEBUG":
    v_log_level = 10
elif v_configuration["LOGGING"].upper() == "INFO":
    v_log_level = 20
elif v_configuration["LOGGING"].upper() == "WARNING":
    v_log_level = 30
elif v_configuration["LOGGING"].upper() == "ERROR":
    v_log_level = 40
elif v_configuration["LOGGING"].upper() == "CRITICAL":
    v_log_level = 50
coloredlogs.install(level=v_configuration["LOGGING"])
v_logger.setLevel(v_log_level)

if v_configuration.get("PREFIX"):
    v_prefix = v_configuration["PREFIX"]

if not v_configuration.get("TITLE"):
    v_configuration["TITLE"] = ""


# ==================================================================
# parsing of main configuration file
# and command line arguments END
# ==================================================================

# ==================================================================
# dynamically import modules BEGIN
# ==================================================================
lst = os.listdir("plugins")
#for file in lst:
#    res = {}
#    file_path = os.path.abspath("plugins") + os.sep + file
#    logging.debug("file_path is [" + file_path + "]")
#    if os.path.isdir(file_path) and os.path.exists(file_path + os.sep + "__init__.py"):
#        logging.debug("trying to append module files")
#        v_files.append()
    # load the modules
#    for module_file in v_files:
#        res[module_file] = __import__("plugins." + module_file, fromlist=["*"])
#logging.debug("files in plugins/")
#logging.debug(v_files)

for file in lst:
    if "__" not in file and file != "plugin_xlsx.py" and file != "plugin_report_conf.py" and file.endswith(".py"):
        # module_name = file[:-3]
        # import_module("plugins." + module_name, '*')
         module_name = file.split('.')[0]
         globals()[module_name] = import_module('plugins.' + module_name, '*')

modules = [m for m in sys.modules.values() if m.__name__.startswith('plugins.plugin_')]
# ==================================================================
# dynamically import modules END
# ==================================================================


# creating xls file
if len(v_configuration["TITLE"]) == 0:
    v_workbook = xlsxwriter.Workbook(v_prefix +
                                     "report_" + date.today().strftime("%Y%m%d") + ".xlsx",  {'constant_memory': True})
else:
    v_workbook = xlsxwriter.Workbook(v_prefix +
                                     v_configuration["TITLE"] + "_" + date.today().strftime("%Y%m%d") +
                                     ".xlsx",  {'constant_memory': True})
if not v_configuration.get("AUTHOR"):
    v_conf_author = 'unknown'
else:
    v_conf_author = v_configuration["AUTHOR"]
if not v_configuration.get("COMPANY"):
    v_conf_company = 'unknown'
else:
    v_conf_company = v_configuration["COMPANY"]
v_workbook.set_properties({
    'title': v_configuration["TITLE"],
    'subject': 'awr based charts for ' + v_configuration["TITLE"],
    'author': v_conf_author,
    'manager': '',
    'company': v_conf_company,
    'category': 'awr data',
    'keywords': 'awr, perfomance, analyze',
    'created':  date.today(),
    'comments': 'Created with cx_Oracle and xlsxwriter'})
header_format = v_workbook.add_format({'bg_color': '#FFBFBF',
                                       'align': 'center',
                                       'font_name': 'Arial',
                                       'font_size': 10})
integer_format = v_workbook.add_format({'num_format': '#,###0.0',
                                        'align': 'right',
                                        'font_name': 'Arial',
                                        'font_size': 10})
float_format = v_workbook.add_format({'num_format': '#,###0.0000',
                                      'align': 'right',
                                      'font_name': 'Arial',
                                      'font_size': 10})
if len(v_configuration["TITLE"]) == 0:
    v_logger.info('Created xls report [' + v_prefix +
                   "report_" + date.today().strftime("%Y%m%d") + ".xls]")
else:
    v_logger.info('Created xls report [' + v_prefix +
                   v_configuration["TITLE"] + "_" + date.today().strftime("%Y%m%d") +
                   ".xls]")

try:
    # open report configuration file
    if sys.version_info.major >= 3:
        v_config_file = open(v_configuration["REPORT_CONF"], encoding="utf_8")
    if sys.version_info.major == 2:
        v_config_file = open(v_configuration["REPORT_CONF"])
except OSError as err:
    v_logger.error("OS error: {0}".format(err))
    raise
except Exception:
    v_logger.error("Unexpected error:", sys.exc_info()[0])
    raise

for v_line in v_config_file.readlines():
    if not v_line.startswith("\n") and not v_line.startswith("#"):
        v = v_line.replace("\r", "").replace("\n", "")
        v_conf.append(v.split(":"))

v_config_file.close()
v_logger.info('Successfully read report config file ' + v_configuration["REPORT_CONF"])

v_conf = sorted(v_conf, key=lambda x: int(x[0]), reverse=False)

# ==================================================================
# for each query in report configuration file
for config_line in v_conf:
    v_logger.info('Processing plugin [' + config_line[1]
                  + '] with script file [' + config_line[2]
                  + '] for id [' + str(config_line[0]) + ']')

    v_worksheet = v_workbook.add_worksheet(config_line[3])

    v_id,\
    v_plugin,\
    v_source_file,\
    v_worksheet_title,\
    v_chart_title,\
    v_chart_conf_file,\
    v_columns_conf_file = parse_report_conf_line(v_configuration, v_logger, config_line, v_workbook, v_worksheet)

    v_logger.debug("Current config_line is:")
    v_logger.debug("v_id=[" + v_id + "]")
    v_logger.debug("v_plugin=[" + v_plugin + "]")
    v_logger.debug("v_source_file=[" + v_source_file + "]")
    v_logger.debug("v_worksheet_title=[" + v_worksheet_title + "]")
    v_logger.debug("v_chart_title=[" + v_chart_conf_file + "]")
    v_logger.debug("v_columns_conf_file=[" + v_columns_conf_file + "]")

    # v_title, v_data = plugin_oradb.oradb_gather_data(v_configuration, v_logger, config_line, v_workbook, v_worksheet)
    # get data from plugin using source file - dynamic call function from dynamic imported modules
    for module in modules:
        if module.__name__ == 'plugins.plugin_%s' % v_plugin:
            gather_data = getattr(module, "%s_gather_data" % v_plugin)
            v_title, v_data = gather_data(v_configuration, v_logger, config_line, v_workbook, v_worksheet)

    # ==================================================================
    # if we not use calculating columns, write data to worksheet
    # if v_columns_conf == "":
    if v_columns_conf_file == "default":
        v_logger.debug("id [" + v_id + "] is default data inserting")
        default_write_to_the_worksheet(v_configuration, v_logger, config_line, v_workbook, v_worksheet,
                                       v_title,
                                       v_data,
                                       header_format,
                                       integer_format,
                                       float_format)
    # adding calculating columns
    # if v_columns_conf != "":
    else:
        v_logger.debug("id [" + v_id + "] is custom data inserting")
        custom_write_to_the_worksheet(v_configuration, v_logger, config_line, v_workbook, v_worksheet,
                                      v_title,
                                      v_data,
                                      header_format,
                                      integer_format,
                                      float_format)

    # ==================================================================
    # Chart printing

    # If default, use all data
    if v_chart_conf_file == "default":
        v_logger.debug("id [" + v_id + "] is default charts creating")
        add_default_chart_to_xlsx(v_configuration, v_logger, config_line, v_workbook, v_worksheet,
                                  v_title,
                                  len(v_data))

        v_logger.debug("Created default chars to this data")
    # If not default, use config file
    elif v_chart_conf_file != "none":
        v_logger.debug("id [" + v_id + "] is custom charts creating")
        add_custom_charts_to_xlsx(v_configuration, v_logger, config_line, v_workbook, v_worksheet,
                                  len(v_data))

        v_logger.debug("Created custom charts to this data")
    # if none, do not print charts
    elif v_chart_conf_file == "none":
        v_logger.debug("Don't creating charts for this data")

v_logger.debug("Closing workbook")
v_workbook.close()
exit()


plugins/plugin_report_conf.py
# -*- coding: utf-8 -*-
import sys


def parse_report_conf_line(l_configuration, l_logger, l_config_line, l_workbook, l_worksheet):
    """

    :param l_configuration: runtime configuration gathered from general.conf
    :param l_logger: runtime logger
    :param l_config_line: current report configuration line from REPORT_CONF
    :param l_workbook: current workbook
    :param l_worksheet: current worksheet
    :return: parsed to variables l_config_line
    """
    # id:plugin:script file:worksheet title:название графика:конф отрисовки графика:конфиг для вычисляемых столбцов
    #  0:     1:          2:              3:               4:                     5:                              6
    l_id = l_config_line[0]
    l_plugin = l_config_line[1]
    l_source_file = l_config_line[2]
    l_worksheet_title = l_config_line[3]
    if l_config_line[3] == "":
        l_chart_title = l_configuration["TITLE"] + " " + "не именованный график"
    else:
        l_chart_title = l_configuration["TITLE"] + " " + l_config_line[4]
    l_chart_conf_file = l_config_line[5]
    l_columns_conf_file = l_config_line[6]

    if sys.version_info.major == 2:
        l_chart_title = l_chart_title.decode('utf8')

    return l_id, l_plugin, l_source_file, l_worksheet_title, l_chart_title, l_chart_conf_file, l_columns_conf_file


ssh/atop_parser.sh
#!/usr/bin/env bash
# parser for atop daemon logs, running like this:
# 		/usr/bin/atop -a -w /var/log/atop/atop_20190417 600 

################################################################################################################################################
 #      Colored output variables
################################################################################################################################################
COLOR_NC='\e[0m' # No Color
COLOR_WHITE='\e[1;37m'
COLOR_BLACK='\e[0;30m'
COLOR_BLUE='\e[0;34m'
COLOR_LIGHT_BLUE='\e[1;34m'
COLOR_GREEN='\e[0;32m'
COLOR_LIGHT_GREEN='\e[1;32m'
COLOR_CYAN='\e[0;36m'
COLOR_LIGHT_CYAN='\e[1;36m'
COLOR_RED='\e[0;31m'
COLOR_LIGHT_RED='\e[1;31m'
COLOR_PURPLE='\e[0;35m'
COLOR_LIGHT_PURPLE='\e[1;35m'
COLOR_BROWN='\e[0;33m'
COLOR_YELLOW='\e[1;33m'
COLOR_GRAY='\e[0;30m'
COLOR_LIGHT_GRAY='\e[0;37m'

################################################################################################################################################
 #      Default variables
################################################################################################################################################
CURRENT_PID=$$
START_TIME=`date +%Y%m%d_%H%M`
LOG_LEVEL=30

DSK=0
LVM=0
CPU=0
SWAP=0
#PAG=0
RAM=0
USERPARSER=0
declare -a DSK_NAMES
declare -a DSK_LOGS
declare -a LVM_NAMES
declare -a LVM_LOGS

################################################################################################################################################
 #	Output functions
################################################################################################################################################
echo_error() {
        LOG_LEVEL=$1
        shift
        if [ $LOG_LEVEL -le 40 ]
        then
                echo -e "  ERROR `date '+%y.%m.%d %H:%M:%S'`: ${COLOR_RED}$@${COLOR_NC}";
        fi
}
echo_warn()  {
        LOG_LEVEL=$1
        shift
        if [ $LOG_LEVEL -le 30 ]
        then
                echo -e "WARNING `date '+%y.%m.%d %H:%M:%S'`: ${COLOR_YELLOW}$@${COLOR_NC}";
        fi
}
echo_info()  {
        LOG_LEVEL=$1
        shift
        if [ $LOG_LEVEL -le 20 ]
        then
                echo -e "   INFO `date '+%y.%m.%d %H:%M:%S'`: ${COLOR_GREEN}$@${COLOR_NC}";
        fi
}
echo_debug() {
        LOG_LEVEL=$1
        shift
        if [ $LOG_LEVEL -le 10 ]
        then
                echo -e "  DEBUG `date '+%y.%m.%d %H:%M:%S'`: ${COLOR_CYAN}$@${COLOR_NC}";
        fi
}

echo_usage() {
        echo "Usage: `basename $0` [options]
                --help  -h                              show this help
		--loglevel	level			Logging level
								Default: WARNING
								Can be:
								 SILENT
								 ERROR
								 WARNING
								 INFO
								 DEBUG
		--dsk		block device name	Parse atop logs for block device
		--lvm		lvol name			Parse atop logs for logical volume
		--cpu					Parse atop logs for CPU
		--swap					Parse atop logs for swap
		--ram					Parse atop logs for RAM
		--user		username		Parse atop logs for RAM/CPU usage of user username
	Note: supported only RHEL6/OEL6 (RHEL7/OEL7 not tested; RHEL5/OEL5 not supported; Debian not supported)
"
#		--pag					Parse atop logs for PAG (paging frequency)
}

################################################################################################################################################
 #	Atop logs parsing functions
################################################################################################################################################
atop_parse_device() {
	local DEVICE=$1
	local ATOP_FILE=$2
	local TEMP_FILE=$3
	echo_debug $LOG_LEVEL "Processing ${ATOP_FILE} for ${DEVICE}"
	/usr/bin/atop -r $ATOP_FILE -P $DEVICE | awk '{if ( $6 == 600 ) print $0;}' | awk '{print $3" "$5}' | sort -u > $TEMP_FILE
}

################################################################################################################################################
 #	Parsing command line arguments
################################################################################################################################################
if [ $# -eq 0 ]
then
	echo_usage
	exit 1
fi

while [ "$1" != "" ]
do
	case "$1" in
		"-h"|"--help")
			echo_usage
			exit 0
		;;
		"--loglevel")
			case "$2" in
				"SILENT2")
					LOG_LEVEL=70
				;;
				"SILENT")
					LOG_LEVEL=60
				;;
				"ERROR")
					LOG_LEVEL=40
				;;
                                "WARNING")
                                        LOG_LEVEL=30
                                ;;
                                "INFO")
                                        LOG_LEVEL=20
                                ;;
                                "DEBUG")
                                        LOG_LEVEL=10
                                ;;
                                *)
                                        LOG_LEVEL=30
                                ;;
	                esac
                        shift 2
		;;		
		"--dsk")
			DSK=1
			DSK_NAMES+=("$2")
			DSK_LOGS+=("/tmp/atop_parser_DSK_$2_${START_TIME}_${CURRENT_PID}.log")
			LAST_DSK=${#DSK_NAMES[@]}
			LAST_DSK=$((LAST_DSK - 1 ))
#			echo "unixtime device read write avq avio" > ${DSK_LOGS[$LAST_DSK]}
			shift 2
		;;
		"--lvm")
			LVM=1
			LVM_NAMES+=("$2")
			LVM_LOGS+=("/tmp/atop_parser_LVM_$2_${START_TIME}_${CURRENT_PID}.log")
			LAST_LVM=${#LVM_NAMES[@]}
			LAST_LVM=$((LAST_LVM - 1 ))
#			echo "unixtime device read write avq avio" > ${LVM_LOGS[$LAST_LVM]}
			shift 2
		;;
		"--cpu")
			CPU=1
			CPU_LOG="/tmp/atop_parser_CPU_${START_TIME}_${CURRENT_PID}.log"
			shift
		;;
		"--swap")
			SWAP=1
			SWAP_LOG="/tmp/atop_parser_SWP_${START_TIME}_${CURRENT_PID}.log"
			shift
		;;
# there is no PAG on rhel6 with atop-1.26-1.el6.2.x86_64
#		"--pag")
#			PAG=1
#			PAG_LOG="/tmp/atop_parser_PAG_${START_TIME}_${CURRENT_PID}.log"
#			echo "unixtime" > $PAG_LOG
#			echo_warn $LOG_LEVEL "Log file for PAG is [$PAG_LOG]"
#			shift
#		;;
		"--ram")
			RAM=1
			RAM_LOG="/tmp/atop_parser_MEM_${START_TIME}_${CURRENT_PID}.log"
			shift
		;;
		"--user")
			USERPARSER=1
			USERNAME=$2
			USER_LOG="/tmp/atop_parser_USER_${START_TIME}_${CURRENT_PID}.log"
			shift 2
		;;
		*)
			echo_error $LOG_LEVEL "Unknown parameter [$1]"
			echo_usage
			exit 1
		;;
	esac
done


if [ $LOG_LEVEL -lt 70 ]
then
	if [ $DSK -eq 1 ]
	then
		for i in ${!DSK_NAMES[@]}
		do
			echo "unixtime read write avq avio" > ${DSK_LOGS[$LAST_DSK]}
			echo_warn $LOG_LEVEL "Log file for DSK $2 is [${DSK_LOGS[$LAST_DSK]}]"
		done
	fi
	if [ $LVM -eq 1 ]
	then
		for i in ${!LVM_NAMES[@]}
		do
			echo "unixtime read write avq avio" > ${LVM_LOGS[$LAST_DSK]}
			echo_warn $LOG_LEVEL "Log file for LVM $2 is [${LVM_LOGS[$LAST_DSK]}]"
		done
	fi
	if [ $CPU -eq 1 ]
	then
		echo "unixtime sys user irq idle wait" > $CPU_LOG
		echo_warn $LOG_LEVEL "Log file for CPU is [$CPU_LOG]"
	fi
	if [ $RAM -eq 1 ]
	then
		echo "unixtime free cache buff slab" > $RAM_LOG
		echo_warn $LOG_LEVEL "Log file for RAM is [$RAM_LOG]"
	fi
	if [ $SWAP -eq 1 ]
	then
		echo "unixtime total free vmcom vmlim" > $SWAP_LOG
		echo_warn $LOG_LEVEL "Log file for SWAP is [$SWAP_LOG]"
	fi
fi


################################################################################################################################################
 #	Parsing log files
################################################################################################################################################


for file in /var/log/atop/atop_*; do
	[ -e "$file" ] || continue
	echo_debug $LOG_LEVEL "parsing [$file]"
	if [ $USERPARSER -eq 1 ]
	then
		DEVICE="CPU"
		TEMP_FILE="/tmp/atop_parser_${DEVICE}_${START_TIME}_${CURRENT_PID}.temp"
		USER_TEMP_LOG="${TEMP_FILE}.log"
		atop_parse_device $DEVICE $file $TEMP_FILE
		while read -r LINE
		do
			v_unixtime=`echo $LINE | awk '{print $1}'`
			v_time=`echo $LINE | awk '{print $2}'`
			/usr/bin/atop -l -r $file -b $v_time -L 160 -u | head -n 40 |  grep $USERNAME | awk -v ut=$v_unixtime ' { print ut" "$4" "$5" "$10 } ' | sed 's/%//g' >> $USER_TEMP_LOG
		done < $TEMP_FILE
		rm -f $TEMP_FILE
		awk ' {
if ( $2 ~ '/G/' ) {
    gsub("G","",$2)
    $2=$2*1024*1024/4
}
if ( $2 ~ '/M/' ) {
    gsub("M","",$2)
    $2=$2*1024/4
}
if ( $3 ~ '/G/' ) {
    gsub("G","",$3)
    $3=$3*1024*1024/4
}
if ( $3 ~ '/M/' ) {
    gsub("M","",$3)
    $3=$3*1024/4
}
printf "%s %8.1f %8.1f %8.1f\n", $1, $2, $3, $4 } ' $USER_TEMP_LOG >> $USER_LOG
		rm -f $USER_TEMP_LOG
	fi
	if [ $DSK -eq 1 ]
	then
		DEVICE="DSK"
		TEMP_FILE="/tmp/atop_parser_${DEVICE}_${START_TIME}_${CURRENT_PID}.temp"
		atop_parse_device $DEVICE $file $TEMP_FILE
		for i in ${!DSK_NAMES[@]}
		do
			DEVICE_NAME=${DSK_NAMES[$i]}
			while read -r LINE
			do
				v_unixtime=`echo $LINE | awk '{print $1}'`
				v_time=`echo $LINE | awk '{print $2}'`
#				echo "q" | /usr/bin/atop -l -r $file -b $v_time -L 160 | head -n 20 | egrep "^DSK .*" | egrep "$DEVICE_NAME" | awk -F "|" '{print $2" "$4" "$5" "$10" "$11}' | awk -v ut=$v_unixtime '{print ut" "$1" "$3" "$5" "$7" "$9}' >> ${DSK_LOGS[$i]}
				echo "q" | /usr/bin/atop -l -r $file -b $v_time -L 160 | head -n 20 | egrep "^DSK .*" | egrep "$DEVICE_NAME" | awk -F "|" '{print $2" "$4" "$5" "$10" "$11}' | awk -v ut=$v_unixtime '{print ut" "$3" "$5" "$7" "$9}' >> ${DSK_LOGS[$i]}
			done < $TEMP_FILE
		done
		rm -f $TEMP_FILE
	fi
	if [ $LVM -eq 1 ]
	then
		DEVICE="LVM"
		TEMP_FILE="/tmp/atop_parser_${DEVICE}_${START_TIME}_${CURRENT_PID}.temp"
		atop_parse_device $DEVICE $file $TEMP_FILE
		for i in ${!LVM_NAMES[@]}
		do
			DEVICE_NAME=${LVM_NAMES[$i]}
			while read -r LINE
			do
				v_unixtime=`echo $LINE | awk '{print $1}'`
				v_time=`echo $LINE | awk '{print $2}'`
#				echo "q" | /usr/bin/atop -l -r $file -b $v_time -L 160 | head -n 20 | egrep "^LVM .*" | egrep "$DEVICE_NAME" | awk -F "|" '{print $2" "$4" "$5" "$10" "$11}' | awk -v ut=$v_unixtime '{print ut" "$1" "$3" "$5" "$7" "$9}' >> ${LVM_LOGS[$i]}
				echo "q" | /usr/bin/atop -l -r $file -b $v_time -L 160 | head -n 20 | egrep "^LVM .*" | egrep "$DEVICE_NAME" | awk -F "|" '{print $2" "$4" "$5" "$10" "$11}' | awk -v ut=$v_unixtime '{print ut" "$3" "$5" "$7" "$9}' >> ${LVM_LOGS[$i]}
			done < $TEMP_FILE
		done
		rm -f $TEMP_FILE
	fi
	if [ $CPU -eq 1 ]
	then
		DEVICE="CPU"
		TEMP_FILE="/tmp/atop_parser_${DEVICE}_${START_TIME}_${CURRENT_PID}.temp"
		atop_parse_device $DEVICE $file $TEMP_FILE
		while read -r LINE
		do
			v_unixtime=`echo $LINE | awk '{print $1}'`
			v_time=`echo $LINE | awk '{print $2}'`
			echo "q" | /usr/bin/atop -l -r $file -b $v_time -L 160 | head -n 20 | egrep "^CPU." | awk -F "|" -v ut=$v_unixtime ' { print ut" "$2" "$3" "$4" "$5" "$6 } ' | awk ' { print $1, $3, $5, $7, $9, $11 } ' | sed 's/%//g' >> $CPU_LOG
		done < $TEMP_FILE
		rm -f $TEMP_FILE
	fi

	if [ $SWAP -eq 1 ]
	then
		DEVICE="SWP"
		TEMP_FILE="/tmp/atop_parser_${DEVICE}_${START_TIME}_${CURRENT_PID}.temp"
		SWP_TEMP_LOG="${TEMP_FILE}.log"
		atop_parse_device $DEVICE $file $TEMP_FILE
		while read -r LINE
		do
			v_unixtime=`echo $LINE | awk '{print $1}'`
			v_time=`echo $LINE | awk '{print $2}'`
			# echo "q" | /usr/bin/atop -l -r $file -b $v_time -L 160 | head -n 20 | egrep "^SWP." |  awk -F "|" ' { print $2, $3, $10, $11 } ' | awk -v ut=$v_unixtime '{print ut, $2, $4, $6, $8 } ' | sed 's/G//g' >> $SWP_TEMP_LOG
			echo "q" | /usr/bin/atop -l -r $file -b $v_time -L 160 | head -n 20 | egrep "^SWP." |  awk -F "|" ' { print $2, $3, $10, $11 } ' | awk -v ut=$v_unixtime '{print ut, $2, $4, $6, $8 } ' >> $SWP_TEMP_LOG
		done < $TEMP_FILE
				awk ' {
if ( $2 ~ '/G/' ) {
    gsub("G","",$2)
    $2=$2*1024*1024/4
}
if ( $2 ~ '/M/' ) {
    gsub("M","",$2)
    $2=$2*1024/4
}
if ( $3 ~ '/G/' ) {
    gsub("G","",$3)
    $3=$3*1024*1024/4
}
if ( $3 ~ '/M/' ) {
    gsub("M","",$3)
    $3=$3*1024/4
}
if ( $4 ~ '/G/' ) {
    gsub("G","",$4)
    $4=$4*1024*1024/4
}
if ( $4 ~ '/M/' ) {
    gsub("M","",$4)
    $4=$4*1024/4
}
if ( $5 ~ '/G/' ) {
    gsub("G","",$5)
    $5=$5*1024*1024/4
}
if ( $5 ~ '/M/' ) {
    gsub("M","",$5)
    $5=$5*1024/4
}
printf "%s %8.1f %8.1f %8.1f %8.1f\n", $1, $2, $3, $4, $5 } ' $SWP_TEMP_LOG >> $SWAP_LOG
		rm -f $SWP_TEMP_LOG
		rm -f $TEMP_FILE
	fi

	if [ $RAM -eq 1 ]
	then
		DEVICE="MEM"
		TEMP_FILE="/tmp/atop_parser_${DEVICE}_${START_TIME}_${CURRENT_PID}.temp"
		RAM_TEMP_LOG="${TEMP_FILE}.log"
		atop_parse_device $DEVICE $file $TEMP_FILE
		while read -r LINE
		do
			v_unixtime=`echo $LINE | awk '{print $1}'`
			v_time=`echo $LINE | awk '{print $2}'`
			echo "q" | /usr/bin/atop -l -r $file -b $v_time -L 160 | head -n 20 | egrep "^MEM." | awk -v ut=$v_unixtime '{print ut, $7, $10, $16, $19 } ' >> $RAM_TEMP_LOG
		done < $TEMP_FILE
#$1=strftime("%Y-%m-%d-%H:%M", $1);
		awk ' {
if ( $2 ~ '/G/' ) {
    gsub("G","",$2)
    $2=$2*1024*1024/4
}
if ( $2 ~ '/M/' ) {
    gsub("M","",$2)
    $2=$2*1024/4
}
if ( $3 ~ '/G/' ) {
    gsub("G","",$3)
    $3=$3*1024*1024/4
}
if ( $3 ~ '/M/' ) {
    gsub("M","",$3)
    $3=$3*1024/4
}
if ( $4 ~ '/G/' ) {
    gsub("G","",$4)
    $4=$4*1024*1024/4
}
if ( $4 ~ '/M/' ) {
    gsub("M","",$4)
    $4=$4*1024/4
}
if ( $5 ~ '/G/' ) {
    gsub("G","",$5)
    $5=$5*1024*1024/4
}
if ( $5 ~ '/M/' ) {
    gsub("M","",$5)
    $5=$5*1024/4
}
printf "%s %8.1f %8.1f %8.1f %8.1f\n", $1, $2, $3, $4, $5 } ' $RAM_TEMP_LOG >> $RAM_LOG
		rm -f $RAM_TEMP_LOG
		rm -f $TEMP_FILE
	fi
done

if [ $LOG_LEVEL -eq 70 ]
then
	if [ $DSK -eq 1 ]
	then
		for i in ${!DSK_NAMES[@]}
		do
			echo "${DSK_LOGS[$LAST_DSK]} unixtime read write avq avio"
		done
	fi
	if [ $LVM -eq 1 ]
	then
		for i in ${!LVM_NAMES[@]}
		do
			echo "${LVM_LOGS[$LAST_LVM]} unixtime read write avq avio"
		done
	fi
	if [ $CPU -eq 1 ]
	then
		echo "$CPU_LOG unixtime sys user irq idle wait"
	fi
	if [ $RAM -eq 1 ]
	then
		echo "$RAM_LOG unixtime free cache buff slab"
	fi
	if [ $SWAP -eq 1 ]
	then
		echo "$SWAP_LOG unixtime total free vmcom vmlim"
	fi
	if [ $USERPARSER -eq 1 ]
	then
		echo "$USER_LOG unixtime vsize rsize cpu"
	fi
fi

exit 0

